### PR TITLE
Add a report for invalid ISO8601 dates

### LINKED ISSUE #4072: Report on invalid ISO8601 dates in Cocina description
This is to check remediation progress for invalid ISO 8601 dates. For valid patterns, see #3849.

Desired output: druid, collection druid, catkey, date value(s).

Check for values in description..date..value with description..date.encoding.code iso8601.

### PR DESCRIPTION
**Pull request by jcoyne**
## Why was this change made? 🤔
Fixes #4072

## How was this change tested? 🤨
locally

### CODE REVIEW

File: app/reports/bad_iso8601_date.rb [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+      json = dro.description.to_json
+      date_values = path.on(json)
+      date_values.each do |value|
+        next if regex.match?(value)
+
+        puts "#{dro.external_identifier}, #{value}\n" # rubocop:disable Rails/Output
+      end
+    end
+  end
+end
```
Comment (mjgiarlo) [inline app/reports/bad_iso8601_date.rb:19]: Can/should we tell our coverage tools to ignore `app/reports/**/*.rb`? Unless you think we should be writing tests for these reports.

Comment (jcoyne) [reply app/reports/bad_iso8601_date.rb]: Yeah, I think we could ignore them, I just don't know how to do that offhand.

Comment (mjgiarlo) [reply app/reports/bad_iso8601_date.rb]: @jcoyne something something simplecov filter?

Comment (ndushay) [reply app/reports/bad_iso8601_date.rb]: I think it's in rails_helper or in spec_helper, or in some dotfile.

**Comment by ndushay** [comment]
why write tests for a report, instead of tweaking the coverage in spec_helper to ignore reports (this and future reports)?

File: app/validators/iso8601_validator.rb [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+# Supports:
+#  2015-06-27
+#  2013-05-09T10:55:29.756-07:00
+#  2013-05-09T10:55:29.756Z
+#  YYYYMMDD
+#  YYYYMMDDThhmm
+#  YYYYMMDDThhmmss
+#  YYYYMMDDhhmm
+#  YYYYMMDDhhmmss
+#  YYYYMMDDhhmmss.sss
```
[the remark this answers is not in the document]
Comment (jcoyne) [reply app/validators/iso8601_validator.rb:14]: I can't think of any good reason not to use DateTime.iso8601. I was unaware of it's existence.

[the remark this answers is not in the document]
Comment (mjgiarlo) [reply app/validators/iso8601_validator.rb:14]: @jcoyne OK, great!

**Review by mjgiarlo — APPROVED** [review]

**Pull request merged**